### PR TITLE
[WIP]Speed up GPU operations by 2x if involved in scalar operations alot.

### PR DESCRIPTION
**Pull request by kartik4949**
So PR, checks if tensor has no requires_grad then it wont spend costly operations on calculating the gradient for it.

### CODE REVIEW

**Comment by kartik4949** [comment]
e.g 
```
from  tinygrad import tensor
import numpy as np

import time
st  =  time.time()
for _ in range(100):
    x = tensor.Tensor([np.random.uniform(size =(100,100))], gpu=True)
    y = tensor.Tensor([np.random.uniform(size =(100,100))], gpu=True)

    yz = tensor.Tensor([np.random.uniform(size =(100,100))], gpu=True, requires_grad=False)
    yz1 = tensor.Tensor([np.random.uniform(size =(100,100))], gpu=True)
    y1 = tensor.Tensor([np.random.uniform(size =(100,100))], gpu=True)
    y2 = tensor.Tensor([np.random.uniform(size =(100,100))], gpu=True)
    y3 = tensor.Tensor([np.random.uniform(size =(100,100))], gpu=True)
    y4 = tensor.Tensor([np.random.uniform(size =(100,100))], gpu=True)
    z = x * y
    z1 = y1 * y2 * 100.0
    z2 = y3 * y4 * 100.0
    out = z1 + z2 + yz*yz1 + 1000.0
    z = out.mean()
    z.backward()
print(time.time() - st)
```
without the require_grad check 
took : 4.1s

with the require_grad check
took : 1.5s



will refactor the pr....

**Comment by kartik4949** [comment]
its weird why gpu is not found in test check 
can u look into @geohot

**Comment by kartik4949** [comment]
1. today's  pytest checks
all of them finished in 29s no CUDA found 


and
 
commit f50dcc12ac57d5c1212840660697c77820bba4f0
broke gpu results.

**Comment by geohot** [comment]
I'm not going to refactor this or the slice PR. You have to make them clean enough if you want them merged, I'm def not merging anything with requires grad checks in the ops.

**Comment by kartik4949** [comment]
> I'm not going to refactor this or the slice PR. You have to make them clean enough if you want them merged, I'm def not merging anything with requires grad checks in the ops.

def I will be refactoring 'my' pr's.

**Comment by kartik4949** [comment]
closing the PR, couldnt find any other way of checking requires_grad, 'tried with def backward too'
may be someone can implement better.
